README.md
# TouchID Remote - use touchID over SSH!

This project builds on the awesome [touch2sudo](https://github.com/prbinu/touch2sudo) and allows it to be used with
remote ssh connections without using SSH agent forwarding which can introduce security vulnerabilities.


## Note

After compiling this I discovered a [simpler way](https://m4.rkw.io/blog/touchid-over-ssh-part-2-secretive-agent.html) to do it that might work
better for some people.

## Overview

There are two components:

1) A PAM module which must be installed on the remote server ([pam\_tid.so](https://github.com/m4rkw/touchid-remote/blob/master/pam_tid.c))

2) A python authentication daemon which runs locally and invokes touch2sudo
([server.py](https://github.com/m4rkw/touchid-remote/blob/master/server.py))

The basic process is:

1) When the user connects to the remote machine via ssh, a tunnel is created
back to the authentication server on the client machine.

2) When the user invokes sudo the PAM module executes and checks whether the
authentication service is available and if it's allowed to be used with the
current user. If these conditions are met and it can connect to the
authentication service through the remote tunnel it will pass a static key to
the authentication server.

3) The authentication server on the client machine validates the shared secret,
if it's invalid then it returns '0' and the PAM module returns an error status
which then causes PAM to defer to the next authentication mechanism (which would
typically be a password prompt).

4) If the shared secret is correct the authentication server invokes touch2sudo
which prompts the user to authenticate with TouchID. If this is successful this
is indicated back to the remote machine with a second shared secret.

5) The PAM module validates the second shared secret and if this is correct then
the user is authenticated and the sudo command completes.

Additionally the USE\_TOUCHID environment variable must be set to '1' in order
for TouchID to be used. This is so that in a scenario where you have a mac
connected (so the tunnel is available) but you're connecting on a different
device that doesn't have a TouchID authentication server (eg an iPhone) you can
still authenticate normally.

## Installation

1) Compile and install the PAM module on your remote servers:

````
$ sudo apt install libconfig-dev
$ git clone https://github.com/m4rkw/touchid-remote
$ cd touchid-remote
$ make
$ sudo make install
````

Note: this must be installed as the user you normally use the mac with, not as
root.

2) Grab touch2sudo from: https://github.com/prbinu/touch2sudo/releases and
install it somewhere in your PATH, eg /usr/local/bin/touch2sudo

3) Install the authentication server on your mac terminal:

````
$ git clone https://github.com/m4rkw/touchid-remote
$ make install
````

4) Edit ~/.touchid-remote/touchid-remote.conf and set key1 and key2 to long
random strings (max 128 chars). Also edit /etc/touchid-remote.conf on the remote
machine(s) and set key1 and key2 to the same values. On the mac make sure that
touch2sudo\_path is correct.

5) Start the authentication service on the mac:

````
$ launchctl load -w ~/Library/LaunchAgents/com.m4rkw.touchid-remote.plist
````

The -w flag indicates that this should run persistently, ie it will be loaded
when the system restarts.

6) Test that the authentication server works:

````
$ curl http://localhost:61111 -H "Auth: key1"
````

Replace "key1" with your configured secret for key1. When you execute this
command you should see a touch2sudo prompt, if this is successfully validated
the output of the curl command will be the value of key2.

7) Add this auth line to the top of your /etc/pam.d/sudo on the remote machine(s):

````
#%PAM-1.0
auth sufficient /usr/lib64/security/pam_tid.so
````

8) Allow the USE\_TOUCHID environment variable to be passed from your client
machine to the server by adding it to the AcceptEnv list in
/etc/ssh/sshd\_config:

````
AcceptEnv LANG LC_* USE_TOUCHID
````

Note: restart sshd for this to take effect.

9) Configure the remote tunnel for your ssh connections on the mac, for example
a connection entry in ~/.ssh/config might look like:

````
Host server1
  Hostname server1.mydomain.com
  SendEnv USE_TOUCHID
  RemoteForward localhost:61111 localhost:61111
````

The RemoteForward entry ensures that whenever you connect to this server with
ssh it will attempt to initiate a tunnel back to the authentication server on
the mac client.

Also make sure you have USE\_TOUCHID set to 1:

````
$ echo 'export USE_TOUCHID=1' >> ~/.bashrc
$ source ~/.bashrc
````

Now ssh to the server using this new connection profile, eg:

````
$ ssh server1
````

and then type "sudo bash". You should see a touch2sudo prompt and when this is
successfully authenticated the sudo command will be accepted.

Of course only one such tunnel can exist because only one socket can listen on
localhost:61111 on the remote server, so if you open a second connection you
will see:

````
Warning: remote port forwarding failed for listen port 61111
````

However this doesn't matter and the second connection will still be able to
authenticate with sudo using the tunnel created with the first connection.

## Disclaimer

Use at your own risk, I am not responsible for any security issues that may
arise from using this software.

This project is not affiliated with Apple or the author of touch2sudo in any
way. "TouchID" and "iPhone" are trademarks of Apple.


CHANGELOG.md
# Version history

````
18/02/2022 - 0.0.1 - Initial release
````


server.py
#!/usr/bin/env python3
#
# TouchID Remote server
# https://github.com/m4rkw/touchid-remote

import os
import sys
import signal
import socket
import time
import pwd
import re

CONFIG_FILE = pwd.getpwuid(os.getuid()).pw_dir + '/.touchid-remote/touchid-remote.conf'

class TouchIDRemoteServer:

  def __init__(self):
    self.load_config()


  def load_config(self):
    if not os.path.exists(CONFIG_FILE):
      raise Exception("config file not found: %s" % (CONFIG_FILE))

    self.config = {}

    with open(CONFIG_FILE, 'r') as f:
      for line in f:
        line = line.rstrip()

        match = re.match('^(.*?): (.*)$', line)

        if match:
          if match.group(2).isdigit():
            self.config[match.group(1)] = int(match.group(2))
          else:
            if match.group(2)[0] == '"' and match.group(2)[-1] == '"':
              self.config[match.group(1)] = match.group(2)[1:len(match.group(2))-1]
            else:
              self.config[match.group(1)] = match.group(2)


  def main(self):
    signal.signal(signal.SIGCHLD, signal.SIG_IGN)

    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(("127.0.0.1", self.config['server_port']))
    s.listen(100)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    while 1:
      (cs, addr) = s.accept()
      pid = os.fork()

      if pid == 0:
        s.close()
        self.handle(cs, addr)
        sys.exit(0)

      cs.close()


  def handle(self, cs, addr):
    req = cs.recv(1024).decode()

    resp =  "HTTP/1.1 200 OK\n"
    resp += "Server:localhost\n"
    resp += "Content-type:text/plain\n\n"

    msg = '1'

    if 'Auth:' in req:
      key = req.split('Auth: ')[1].split("\r")[0]

      if key == self.config['key1']:
        rc = os.system(self.config['touch2sudo_path'])

        if rc == 0:
          msg = self.config['key2']

    resp += msg 

    cs.sendall(resp.encode('utf-8'))
    cs.close()


t = TouchIDRemoteServer()
t.main()


pam_tid.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <libconfig.h>
#include <pwd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/stat.h>
#include <security/pam_appl.h>
#include <security/pam_modules.h>

#define CONFIG_FILE "/etc/touchid-remote.conf"

PAM_EXTERN int pam_sm_setcred( pam_handle_t *pamh, int flags, int argc, const char **argv ) {
	return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv) {
	printf("Acct mgmt\n");
	return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_authenticate( pam_handle_t *pamh, int flags,int argc, const char **argv ) {
  FILE *fp;
  char buf[1024];
  int rc;
  struct passwd *p;
  struct sockaddr_in server_address;
  int c, s, r, i, valid_user;
  char *pos;
  struct stat st;
  config_t cfg;
  config_setting_t *setting;
  const char *key1;
  const char *key2;
  const char *user;
  int server_port;
  char *e;

  fp = popen("echo $USE_TOUCHID", "r");
  fread((char *)&buf, 1, 1, fp);
  pclose(fp);

  if (buf[0] != '1') {
    return PAM_AUTH_ERR;
  }

  config_init(&cfg);

  if (!config_read_file(&cfg, CONFIG_FILE)) {
    fprintf(stderr, "%s:%d - %s\n", config_error_file(&cfg), config_error_line(&cfg), config_error_text(&cfg));
    config_destroy(&cfg);
    return PAM_AUTH_ERR;
  }

  if (!config_lookup_int(&cfg, "server_port", &server_port)) {
    fprintf(stderr, "No 'server_port' setting in configuration file.\n");
    return PAM_AUTH_ERR;
  }

  if (!config_lookup_string(&cfg, "key1", &key1)) {
    fprintf(stderr, "No 'key1' setting in configuration file.\n");
    return PAM_AUTH_ERR;
  }

  if (strlen(key1) >128) {
    fprintf(stderr, "Max key length is 128 bytes\n");
    return PAM_AUTH_ERR;
  }

  if (!config_lookup_string(&cfg, "key2", &key2)) {
    fprintf(stderr, "No 'key2' setting in configuration file.\n");
    return PAM_AUTH_ERR;
  }

  if (strlen(key2) >128) {
    fprintf(stderr, "Max key length is 128 bytes\n");
    return PAM_AUTH_ERR;
  }

  setting = config_lookup(&cfg, "users");

  if (setting == NULL) {
    fprintf(stderr, "No 'users' setting in configuration file.\n");
    return PAM_AUTH_ERR;
  }

  if (config_setting_length(setting) == 0) {
    fprintf(stderr, "No users defined in configuration file.\n");
    return PAM_AUTH_ERR;
  }

  p = getpwuid(getuid());

  for(i = 0; i < config_setting_length(setting); i++) {
    config_setting_t *user_setting = config_setting_get_elem(setting, i);

    user = config_setting_get_string(user_setting);

    if (strcmp(user, p->pw_name) == 0) {
      valid_user = 1;
      break;
    }
  }

  if (valid_user == 0) {
    fprintf(stderr, "invalid user\n");
    return PAM_AUTH_ERR;
  }

  s = socket(AF_INET, SOCK_STREAM, 0);

  server_address.sin_family = AF_INET;
  server_address.sin_port = htons(server_port);
  server_address.sin_addr.s_addr = inet_addr("127.0.0.1");

  if (connect(s, (struct sockaddr *)&server_address, sizeof(server_address)) != 0) {
    fprintf(stderr, "error: connect() failed\n");
    return PAM_AUTH_ERR;
  }

  snprintf((char *)&buf, sizeof(buf)-1, "GET /auth HTTP/1.1\rHost: localhost:%d\rUser-Agent: touchid-remote\rAccept: */*\rAuth: %s\r\r", server_port, key1);

  send(s, &buf, strlen(buf), 1);

  memset(&buf, 0, sizeof(buf));

  recv(s, &buf, sizeof(buf)-1, 0);

  close(s);

  for (i=0; i<strlen(buf); i++) {
    if (buf[i] == '\n' && i+2 < strlen(buf) && buf[i+1] == '\n') {
      if (strcmp(&buf[i+2], key2) == 0) {
        return PAM_SUCCESS;
      }
      break;
    }
  }

  return PAM_AUTH_ERR;
}
